File: ClassEval_T/cpp/solution/NumberWordFormatter.cpp

#include <algorithm> 
#include <sstream>   

#include <string>

class NumberWordFormatter {
public:
    NumberWordFormatter();

    std::string format(int x);
    std::string format(double x);
    std::string format(const std::string& x);
    std::string format(std::nullptr_t);
    std::string trans_two(const std::string& s);
    std::string trans_three(const std::string& s);
    std::string parse_more(int i);

private:
    const std::string NUMBER[10];
    const std::string NUMBER_TEEN[10];
    const std::string NUMBER_TEN[9];
    const std::string NUMBER_MORE[4];
    const std::string NUMBER_SUFFIX[20];
};


NumberWordFormatter::NumberWordFormatter()
    : NUMBER{ "", "ONE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN", "EIGHT", "NINE" },
    NUMBER_TEEN{ "TEN", "ELEVEN", "TWELVE", "THIRTEEN", "FOURTEEN", "FIFTEEN", "SIXTEEN", "SEVENTEEN", "EIGHTEEN", "NINETEEN" },
    NUMBER_TEN{ "TEN", "TWENTY", "THIRTY", "FORTY", "FIFTY", "SIXTY", "SEVENTY", "EIGHTY", "NINETY" },
    NUMBER_MORE{ "", "THOUSAND", "MILLION", "BILLION" },
    NUMBER_SUFFIX{ "k", "w", "", "m", "", "", "b", "", "", "t", "", "", "p", "", "", "e" } {}

std::string NumberWordFormatter::format(int x) {
    std::stringstream ss;
    ss << x;
    return format(ss.str());
}

std::string NumberWordFormatter::format(double x) {
    std::stringstream ss;
    ss << x;
    return format(ss.str());
}

std::string NumberWordFormatter::format(const std::string& x) {
    std::string lstr = x.substr(0, x.find('.'));
    std::string rstr;
    auto now = x.find('.');
    if (x.find('.') != std::string::npos)
    {
        rstr = x.substr(x.find('.') + 1);
    }
    std::reverse(lstr.begin(), lstr.end());

    if (lstr.size() % 3 == 1) lstr += "00";
    else if (lstr.size() % 3 == 2) lstr += "0";

    std::string a[5] = { "", "", "", "", "" };
    std::string lm = "";

    for (size_t i = 0; i < lstr.size() / 3; ++i) {
        a[i] = std::string(1, lstr[3 * i + 2]) + std::string(1, lstr[3 * i + 1]) + std::string(1, lstr[3 * i]);
        
        if (a[i] != "000") {
            lm = trans_three(a[i]) + " " + parse_more(i) + " " + lm;
        }
        else {
            lm += trans_three(a[i]);
        }
    }
    std::string xs = !rstr.empty() ? "AND CENTS " + trans_two(rstr) + " " : "";
    if (lm.empty()) {
        return "ZERO ONLY";
    }
    else {
        int nowpos = 0;
        for (int i = lm.size() - 1; i >= 0; --i)
        {
            if (lm[i] != ' ')
            {
                nowpos = i;
                break;
            }
        }
        lm = lm.substr(0, nowpos + 1);
        return lm + " " + xs + "ONLY";
    }
}

std::string NumberWordFormatter::format(std::nullptr_t) {
    return "";
}

std::string NumberWordFormatter::trans_two(const std::string& s) {
    std::string ss = s;
    if (ss.size() == 1) ss = "0" + ss;

    if (ss[0] == '0') {
        return NUMBER[ss[1] - '0'];
    }
    else if (ss[0] == '1') {
        return NUMBER_TEEN[std::stoi(ss) - 10];
    }
    else if (ss[1] == '0') {
        return NUMBER_TEN[ss[0] - '1'];
    }
    else {
        return NUMBER_TEN[ss[0] - '1'] + " " + NUMBER[ss[1] - '0'];
    }
}

std::string NumberWordFormatter::trans_three(const std::string& s) {
    if (s[0] == '0') {
        return trans_two(s.substr(1));
    }
    else if (s.substr(1) == "00") {
        return NUMBER[s[0] - '0'] + " HUNDRED";
    }
    else {
        return NUMBER[s[0] - '0'] + " HUNDRED AND " + trans_two(s.substr(1));
    }
}

std::string NumberWordFormatter::parse_more(int i) {
    return NUMBER_MORE[i];
}

File: ClassEval_T/java/solutuon/NumberWordFormatter.java
package org.example;

public class NumberWordFormatter {
    private final String[] NUMBER = {"", "ONE", "TWO", "THREE", "FOUR", "FIVE", "SIX", "SEVEN", "EIGHT", "NINE"};
    private final String[] NUMBER_TEEN = {"TEN", "ELEVEN", "TWELVE", "THIRTEEN", "FOURTEEN", "FIFTEEN", "SIXTEEN", "SEVENTEEN", "EIGHTEEN", "NINETEEN"};
    private final String[] NUMBER_TEN = {"TEN", "TWENTY", "THIRTY", "FORTY", "FIFTY", "SIXTY", "SEVENTY", "EIGHTY", "NINETY"};
    private final String[] NUMBER_MORE = {"", "THOUSAND", "MILLION", "BILLION"};
    private final String[] NUMBER_SUFFIX = {"k", "w", "", "m", "", "", "b", "", "", "t", "", "", "p", "", "", "e"};

    public String format(Object x) {
        if (x == null) {
            return "";
        }
        return formatString(x.toString());
    }

    public String formatString(String x) {
        String[] parts = x.split("\\.");
        String lstr = parts[0];
        String rstr = parts.length > 1 ? parts[1] : "";
        String lstrrev = new StringBuilder(lstr).reverse().toString();
        String[] a = new String[5];

        if (lstrrev.length() % 3 == 1) {
            lstrrev += "00";
        } else if (lstrrev.length() % 3 == 2) {
            lstrrev += "0";
        }

        StringBuilder lm = new StringBuilder();
        for (int i = 0; i < lstrrev.length() / 3; i++) {
            a[i] = new StringBuilder(lstrrev.substring(3 * i, 3 * i + 3)).reverse().toString();
            if (!a[i].equals("000")) {
                lm.insert(0, transThree(a[i]) + " " + parseMore(i) + " ");
            } else {
                lm.insert(0, transThree(a[i]));
            }
        }

        String xs = !rstr.isEmpty() ? "AND CENTS " + transTwo(rstr) + " " : "";
        if (lm.toString().trim().isEmpty()) {
            return "ZERO ONLY";
        } else {
            return lm.toString().trim() + " " + xs + "ONLY";
        }
    }

    public String transTwo(String s) {
        s = String.format("%2s", s).replace(' ', '0');
        if (s.charAt(0) == '0') {
            return NUMBER[Integer.parseInt(s.substring(1))];
        } else if (s.charAt(0) == '1') {
            return NUMBER_TEEN[Integer.parseInt(s) - 10];
        } else if (s.charAt(1) == '0') {
            return NUMBER_TEN[Integer.parseInt(s.substring(0, 1)) - 1];
        } else {
            return NUMBER_TEN[Integer.parseInt(s.substring(0, 1)) - 1] + " " + NUMBER[Integer.parseInt(s.substring(1))];
        }
    }

    public String transThree(String s) {
        if (s.charAt(0) == '0') {
            return transTwo(s.substring(1));
        } else if (s.substring(1).equals("00")) {
            return NUMBER[Integer.parseInt(s.substring(0, 1))] + " HUNDRED";
        } else {
            return NUMBER[Integer.parseInt(s.substring(0, 1))] + " HUNDRED AND " + transTwo(s.substring(1));
        }
    }

    public String parseMore(int i) {
        return NUMBER_MORE[i];
    }
}